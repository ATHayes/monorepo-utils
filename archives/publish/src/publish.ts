import * as path from "path";

const inquirer = require("inquirer");
const { getPackages } = require("@monorepo-utils/package-utils");
const canNpmPublish = require("can-npm-publish").canNpmPublish;
import { execUnlessDry, log, logPromise } from "./utils/utils";
import { asyncFilter } from "./utils/async-utils";

/**
 * publish packages to npm
 */
export const publish = async ({
    projectDir,
    dry,
    skipPrompt,
    ciMode,
    distTag
}: {
    projectDir: string;
    dry: boolean;
    skipPrompt: boolean;
    ciMode: boolean;
    distTag?: string;
}): Promise<void> => {
    const packages = getPackages(projectDir);
    const publishablePackages = await asyncFilter(packages, async (pkg: any) => {
        try {
            await canNpmPublish(pkg.location);
            return true;
        } catch (error) {
            return false;
        }
    });
    if (publishablePackages.length === 0) {
        return log("No need to publish.");
    }
    // confirm: npm publish if skipPrompt is false
    const answers = skipPrompt
        ? {
              publish: true
          }
        : await inquirer.prompt([
              {
                  type: "confirm",
                  name: "publish",
                  message: `Do you publish these packages?
${publishablePackages
    .map((pkg: any) => {
        let item = `${pkg.packageJSON.name}@${pkg.packageJSON.version}`;
        if (distTag) {
            item += `, distTag: @${distTag}`;
        }
        return `- ${item}`;
    })
    .join("\n")}
`
              }
          ]);
    if (!answers["publish"]) {
        return log("No publish");
    }
    // npm publish
    await publishablePackages.reduce((promise: Promise<any>, pkg: any) => {
        return promise.then(() => {
            const location = pkg.location;
            const newVersion = pkg.packageJSON.version;
            const args: string[] = [];
            // yarn should use --new-version argument for publishing
            // https://github.com/lerna/lerna/issues/896
            const npmPath = process.env.npm_execpath;
            const isYarn = path.basename(npmPath || "npm").startsWith("yarn");
            const bin = isYarn ? "yarn" : "npm";
            if (isYarn) {
                // skip prompt for new version, use existing instead
                // https://yarnpkg.com/en/docs/cli/publish#toc-yarn-publish-new-version
                args.push("--new-version", newVersion, "--non-interactive");
            }
            if (distTag) {
                args.push("--tag", distTag);
            }
            const task = execUnlessDry(`${bin} publish ` + args.join(" "), {
                cwd: location,
                dry
            });
            let message = `${pkg.packageJSON.name}@${pkg.packageJSON.version}`;
            if (distTag) {
                message += `, distTag: @${distTag}`;
            }
            return logPromise(task, message, {
                ciMode: ciMode
            });
        });
    }, Promise.resolve());
};
