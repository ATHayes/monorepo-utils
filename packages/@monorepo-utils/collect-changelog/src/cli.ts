// MIT Â© 2018 azu
"use strict";

import * as fs from "fs";

import * as path from "upath";
import { findChangelog, getChangelog } from "./collect-changelog-from-tag";
import Handlebars = require("handlebars");

export interface ExecuteOptions {
    directory: string;
    changelogFilePath?: string;
    name?: string;
    tag: string;
    templatePath?: string;
}

export async function execute(options: ExecuteOptions) {
    const templatePath = options.templatePath
        ? path.resolve(process.cwd(), options.templatePath)
        : path.join(__dirname, "../template/changelog.handlebars");
    const code = fs.readFileSync(templatePath, "utf-8");
    const template = Handlebars.compile(code);
    if (options.changelogFilePath) {
        const changelog = await getChangelog(options.changelogFilePath, {
            name: options.name ?? "package", // no name
            version: options.tag,
        });
        return template(changelog).trim();
    } else {
        const changelog = await findChangelog(options.directory, options.tag);
        return template(changelog).trim();
    }
}
