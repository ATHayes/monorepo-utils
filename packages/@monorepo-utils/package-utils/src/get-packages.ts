"use strict";

const globby = require("globby");
import loadJsonFile = require("load-json-file");
import * as path from "path";
import * as fs from "fs";

const loadPackage = (packagePath: string) => {
    const pkgJsonPath = path.join(packagePath, "package.json");
    if (fs.existsSync(pkgJsonPath)) {
        return loadJsonFile.sync(pkgJsonPath);
    }
};

const findPackages = (packageSpecs: string[], rootDirectory: string) => {
    return packageSpecs
        .reduce(
            (pkgDirs, pkgGlob) => {
                const target = path.join(rootDirectory, pkgGlob);
                const filePathList = globby.hasMagic(pkgGlob)
                    ? globby.sync(target, {
                          onlyFiles: false
                      })
                    : [target];
                return pkgDirs.concat(filePathList);
            },
            [] as string[]
        )
        .map((location: string) => {
            return { location, package: loadPackage(location) };
        })
        .filter(res => {
            return res.package && res.package.name;
        });
};

export const getPackages = (directory: string) => {
    const lernaJsonPath = path.join(directory, "lerna.json");
    if (fs.existsSync(lernaJsonPath)) {
        const lernaJson = loadJsonFile.sync(lernaJsonPath);
        if (!lernaJson.useWorkspaces) {
            return findPackages(lernaJson.packages, directory);
        }
    }

    const pkgJsonPath = path.join(directory, "package.json");
    if (fs.existsSync(pkgJsonPath)) {
        const pkgJson = loadJsonFile.sync(pkgJsonPath);
        if (pkgJson.workspaces) {
            // "workspaces": []
            if (Array.isArray(pkgJson.workspaces)) {
                return findPackages(pkgJson.workspaces, directory);
            } else if (Array.isArray(pkgJson.workspaces.packages)) {
                // "workspaces": { "packages": [] }
                return findPackages(pkgJson.workspaces.packages, directory);
            }
        }
    }

    // Bail if we don't find any packages
    return [];
};
