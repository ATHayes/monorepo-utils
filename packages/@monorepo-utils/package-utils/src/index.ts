export { getPackages } from "./get-packages";
export type { PackageResult } from "./get-packages";
