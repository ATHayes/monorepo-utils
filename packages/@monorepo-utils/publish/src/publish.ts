const inquirer = require("inquirer");
const { getPackages } = require("@monorepo-utils/package-utils");
const canNpmPublish = require("can-npm-publish").canNpmPublish;
import { execUnlessDry, log, logPromise } from "./utils/utils";
import { asyncFilter } from "./utils/async-utils";

/**
 * publish packages to npm
 */
export const publish = async ({
    projectDir,
    dry,
    skipPrompt,
    ciMode
}: {
    projectDir: string;
    dry: boolean;
    skipPrompt: boolean;
    ciMode: boolean;
}): Promise<void> => {
    const packages = getPackages(projectDir);
    const publishablePackages = await asyncFilter(packages, async (pkg: any) => {
        try {
            await canNpmPublish(pkg.location);
            return true;
        } catch (error) {
            return false;
        }
    });
    if (publishablePackages.length === 0) {
        return log("No need to publish.");
    }
    // confirm: npm publish if skipPrompt is false
    const answers = skipPrompt
        ? {
              publish: true
          }
        : await inquirer.prompt([
              {
                  type: "confirm",
                  name: "publish",
                  message: `Do you publish these packages?
${publishablePackages
                      .map((pkg: any) => {
                          return `- ${pkg.package.name}@${pkg.package.version}`;
                      })
                      .join("\n")}
`
              }
          ]);
    if (!answers["publish"]) {
        return log("No publish");
    }
    // npm publish
    await publishablePackages.reduce((promise: Promise<any>, pkg: any) => {
        return promise.then(() => {
            const location = pkg.location;
            const task = execUnlessDry(`npm publish`, {
                cwd: location,
                dry
            });
            return logPromise(task, `${pkg.package.name}@${pkg.package.version}`, {
                ciMode: ciMode
            });
        });
    }, Promise.resolve());
};
