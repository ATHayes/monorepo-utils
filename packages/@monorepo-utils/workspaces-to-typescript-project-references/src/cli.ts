import meow from "meow";
import path from "path";
import { toProjectReferences } from "./index";

export const cli = meow(
    `
    Usage
      $ workspaces-to-typescript-project-references
 

    Options
      --root             [Path:string] Root directory of the monorepo. 
                         Default: current working directory
      --check            If set the flag, check only differences of tsconfig.json and does not update tsconfig.json.
                         If the check is failed, exit status 1. It is useful for testing.
       
      --plugin           [Path:string] Path to plugin script.
                         Load the plugin script as module and use it. 
                           
    Examples
      # Update project references in tsconfig.json
      $ workspaces-to-typescript-project-references
      # Test on CI
      $ workspaces-to-typescript-project-references --check
`,
    {
        flags: {
            root: {
                type: "string",
                default: process.cwd()
            },
            check: {
                type: "boolean",
                default: false
            },
            plugin: {
                type: "string",
                isMultiple: true
            }
        },
        autoHelp: true,
        autoVersion: true
    }
);

export const run = async (
    _input = cli.input,
    flags = cli.flags
): Promise<{ exitStatus: number; stdout: string | null; stderr: string | null }> => {
    const plugins = Array.isArray(flags.plugin)
        ? flags.plugin.map((pluginPath) => require(path.join(process.cwd(), pluginPath)))
        : undefined;
    const result = toProjectReferences({
        rootDir: flags.root,
        checkOnly: flags.check,
        plugins
    });
    if (result.ok) {
        return {
            exitStatus: 0,
            stdout: flags.check ? "" : "Update Project References!",
            stderr: null
        };
    } else if (result.aggregateError) {
        return {
            exitStatus: 1,
            stdout: null,
            stderr: result.aggregateError.message + "\n\n" + result.aggregateError.errors.join("\n")
        };
    }
    return {
        exitStatus: 0,
        stdout: null,
        stderr: null
    };
};
