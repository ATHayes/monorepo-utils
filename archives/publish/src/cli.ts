#!/usr/bin/env node

import { publish } from "./publish";
import { execRead, getUnexecutedCommands } from "./utils/utils";

const chalk = require("chalk");
const meow = require("meow");

const cli = meow(
    `
    Usage
      $ monorepo-utils-publish

    Options:
      --dry Enable dry run mode
      --yes Yes all
      --ci  CI mode includes --yes flag
      --dist-tag dist-tag for npm

    Examples
      # publish 
      $ monorepo-utils-publish --dry
      # publish with --tag
      # https://docs.npmjs.com/cli/publish
      $ monorepo-utils-publish --dist-tag next
      
      # dry-run mode
      $ monorepo-utils-publish --dry
      # manually publish
      # CI mode
      $ monorepo-utils-publish --ci
`,
    {
        autoVersion: true,
        autoHelp: true,
        flags: {
            dry: {
                type: "boolean"
            },
            yes: {
                type: "boolean"
            },
            ci: {
                type: "boolean"
            },
            distTag: {
                type: "string"
            }
        }
    }
);

const dryRunMode = !!cli.flags.dry;
if (dryRunMode) {
    console.log(
        chalk`
    {green.bold dry-run mode}
    `
    );
}

export async function run() {
    const projectDir = await execRead("git rev-parse --show-toplevel");
    const ciMode = !!cli.flags.ci;
    const distTag = cli.flags.distTag;
    await publish({
        projectDir,
        dry: dryRunMode,
        skipPrompt: ciMode ? true : !!cli.flags.yes,
        ciMode: ciMode,
        distTag
    });
    console.log(
        chalk`
    {green.bold Publish successful!}
    ${getUnexecutedCommands()}
  `.replace(/\n +/g, "\n")
    );
}
