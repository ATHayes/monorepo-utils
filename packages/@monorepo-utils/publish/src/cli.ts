#!/usr/bin/env node

import { publish } from "./publish";
import { execRead, getUnexecutedCommands } from "./utils/utils";

const chalk = require("chalk");
const meow = require("meow");

const cli = meow(
    `
    Usage
      $ monorepo-utils-publish

    Options:
      --dry Enable dry run mode
      --yes Yes all
      --ci  CI mode includes --yes flag

    Examples
      # dry-run mode
      $ monorepo-utils-publish --dry
      # manually publish
      # CI mode
      $ monorepo-utils-publish --ci
`,
    {
        autoVersion: true,
        autoHelp: true,
        flags: {
            dry: {
                type: "boolean"
            },
            yes: {
                type: "boolean"
            },
            ci: {
                type: "boolean"
            }
        }
    }
);

const dryRunMode = !!cli.flags.dry;
if (dryRunMode) {
    console.log(
        chalk`
    {green.bold dry-run mode}
    `
    );
}

export async function run() {
    const projectDir = await execRead("git rev-parse --show-toplevel");
    const ciMode = !!cli.flags.ci;
    await publish({
        projectDir,
        dry: dryRunMode,
        skipPrompt: ciMode ? true : !!cli.flags.yes,
        ciMode: ciMode
    });
    console.log(
        chalk`
    {green.bold Publish successful!}
    ${getUnexecutedCommands()}
  `.replace(/\n +/g, "\n")
    );
}
