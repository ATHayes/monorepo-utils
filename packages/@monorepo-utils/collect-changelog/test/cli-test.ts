import * as path from "path";
import { execute } from "../src/cli";

const independentDir = path.join(__dirname, "fixtures/independent");
const fixedDir = path.join(__dirname, "fixtures/fixed");
describe("cli", () => {
    describe("when fixed mode", () => {
        it("get 0.13.0 from CHANGELOG.md", async () => {
            const actual = await execute({
                directory: fixedDir,
                changelogFilePath: path.join(fixedDir, "CHANGELOG.md"),
                tag: "0.13.0",
            });
            expect(actual).toMatchSnapshot();
        });
    });
    describe("when independent mode", () => {
        it("textlint@10.2.0 should has changelog ", async () => {
            const actual = await execute({
                directory: independentDir,
                tag: "textlint@10.2.0",
            });
            expect(actual).toMatchSnapshot();
        });
        it("@textlint/kernel@2.0.0-next. should has changelog ", async () => {
            const actual = await execute({
                directory: independentDir,
                tag: "@textlint/kernel@2.0.0-next.0",
            });
            expect(actual).toMatchSnapshot();
        });
        it("@textlint/ast-tester@2.0.7 should be empty ", async () => {
            const actual = await execute({
                directory: independentDir,
                tag: "@textlint/ast-tester@2.0.7",
            });
            expect(actual).toBe("");
        });
    });
});
