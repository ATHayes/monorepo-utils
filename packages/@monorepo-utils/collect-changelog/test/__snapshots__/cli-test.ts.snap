// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli @textlint/kernel@2.0.0-next. should has changelog  1`] = `
"## @textlint/kernel@2.0.0-next.0

### fixes

- **textlint:** fix kernel test ([#374](https://github.com/textlint/textlint/issues/374)) ([d6953cc](https://github.com/textlint/textlint/commit/d6953cc))
- **textlint-fixer-formatter:** use paths ([734806f](https://github.com/textlint/textlint/commit/734806f))"
`;

exports[`cli textlint@10.2.0 should has changelog  1`] = `
"## textlint@10.2.0

### fixes

- **textlint:** check textlintrc option exists for internal use ([744da23](https://github.com/textlint/textlint/commit/744da23))

### features

- **textlint:** support --no-textlintrc ([466e257](https://github.com/textlint/textlint/commit/466e257))
- **textlint:** add tests for --no-textlintrc ([289deb3](https://github.com/textlint/textlint/commit/289deb3))
- **textlint:** pass textlintrc options for failure tests ([bc0804d](https://github.com/textlint/textlint/commit/bc0804d))
- **textlint:** simplify textlint --help test ([0d2114d](https://github.com/textlint/textlint/commit/0d2114d))"
`;

exports[`cli when fixed mode get 0.13.0 from CHANGELOG.md 1`] = `
"## project@0.13.0

### fixes

- **docz:** ts warning ([19ccc09](https://github.com/pedronauck/docz/commit/19ccc09))
- **docz-core:** add setMaxListener for chokidar watchers ([6053c16](https://github.com/pedronauck/docz/commit/6053c16))
- **docz-core:** log level based on debug argument ([507e149](https://github.com/pedronauck/docz/commit/507e149))
- **docz-core:** node path resolve modules merge ([27102fd](https://github.com/pedronauck/docz/commit/27102fd))
- **docz-core:** turn off htmlMinifier when loading from templates ([#518](https://github.com/pedronauck/docz/issues/518)) ([9cb0e1d](https://github.com/pedronauck/docz/commit/9cb0e1d))
- **docz-core:** use webpack-dev-server instead of webpack-serve ([4157e05](https://github.com/pedronauck/docz/commit/4157e05))

### features

- **docz-core:** add onCreateWebpackChain hook ([70bb242](https://github.com/pedronauck/docz/commit/70bb242))
- **docz-core:** add promise logger using progress-estimator ([2797608](https://github.com/pedronauck/docz/commit/2797608))
- **docz-core:** use NODE_PATH to resolve modules ([#516](https://github.com/pedronauck/docz/issues/516)) ([cc86f93](https://github.com/pedronauck/docz/commit/cc86f93))
- **docz-theme-default:** set max lines before scroll editor ([#519](https://github.com/pedronauck/docz/issues/519)) ([698261b](https://github.com/pedronauck/docz/commit/698261b))"
`;

exports[`cli when independent mode @textlint/kernel@2.0.0-next. should has changelog  1`] = `
"## @textlint/kernel@2.0.0-next.0

### fixes

- **textlint:** fix kernel test ([#374](https://github.com/textlint/textlint/issues/374)) ([d6953cc](https://github.com/textlint/textlint/commit/d6953cc))
- **textlint-fixer-formatter:** use paths ([734806f](https://github.com/textlint/textlint/commit/734806f))"
`;

exports[`cli when independent mode textlint@10.2.0 should has changelog  1`] = `
"## textlint@10.2.0

### fixes

- **textlint:** check textlintrc option exists for internal use ([744da23](https://github.com/textlint/textlint/commit/744da23))

### features

- **textlint:** support --no-textlintrc ([466e257](https://github.com/textlint/textlint/commit/466e257))
- **textlint:** add tests for --no-textlintrc ([289deb3](https://github.com/textlint/textlint/commit/289deb3))
- **textlint:** pass textlintrc options for failure tests ([bc0804d](https://github.com/textlint/textlint/commit/bc0804d))
- **textlint:** simplify textlint --help test ([0d2114d](https://github.com/textlint/textlint/commit/0d2114d))"
`;
